numpy/random/mtrand/mvhypergeom.c

```diff
@@ -0,0 +1,246 @@
+    rk_hypergeometric.
+
+    Parameters
+    ----------
+    rk_state *state
+        Pointer to a randomkit `rk_state` instance.
+    size_t num_colors
+        The length of the `colors` array.
+    long *colors
+        The array of colors (i.e. the number of each type in the collection
+        from which the random variate is drawn).
+    long nsample
+        The number of objects drawn without replacement for each variate.
+        `nsample` must not exceed sum(colors).  This condition is not checked;
+        it is assumed that the caller has already validated the value.
+    size_t num_sample
+        The number of random variates to generate.  (This is not the size of
+        the array `sample`, because each variate requires `num_colors`
+        values.)
+    long *sample
+        The array that will hold the result.  The length of this array must
+        be the product of `num_colors` and `num_sample`.
+        The array is not initialized; it is expected that array has been
+        initialized with zeros when the function is called.
+
+    Notes
+    -----
+    Here's an example that demonstrates the idea of this algorithm.
+
+    Suppose the urn contains red, green, blue and yellow marbles.
+    Let nred be the number of red marbles, and define the quantities for
+    the other colors similarly.  The total number of marbles is
+
+        total = nred + ngreen + nblue + nyellow.
+
+    To generate a sample using rk_hypergeometric:
+
+       red_sample = rk_hypergeometric(ngood=nred, nbad=total - nred,
+                                      nsample=nsample)
+
+    This gives us the number of red marbles in the sample.  The number of
+    marbles in the sample that are *not* red is nsample - red_sample.
+    To figure out the distribution of those marbles, we again use
+    rk_hypergeometric:
+
+        green_sample = rk_hypergeometric(ngood=ngreen,
+                                         nbad=total - nred - ngreen,
+                                         nsample=nsample - red_sample)
+
+    Similarly,
+
+        blue_sample = rk_hypergeometric(
+                          ngood=nblue,
+                          nbad=total - nred - ngreen - nblue,
+                          nsample=nsample - red_sample - green_sample)
+
+    Finally,
+
+        yellow_sample = total - (red_sample + green_sample + blue_sample).
+
+    The above sequence of steps is implemented as a loop for an arbitrary
+    number of colors in the innermost loop in the code below.  `remaining`
+    is the value passed to `nbad`; it is `total - colors[0]` in the first
+    call to rk_hypergeometric(), and then decreases by `colors[j]` in
+    each iteration.  `num_to_sample` is the `nsample` argument.  It
+    starts at this function's `nsample` input, and is decreased by the
+    result of the call to rk_hypergeometric() in each iteration.
+*/
+
+int multivar_hypergeom_iterative(rk_state *state,
+                                 size_t num_colors, long *colors,
```

Comment:
Is `size_t` the right thing here? The biggest we can get is `npy_intp`.

Reply:
Are there any guarantees about the bit width of `npy_intp` compared to C `long` (signed or unsigned)?  I'm not sure if these can be safely mixed in expressions and comparisons.